File: MODSRC/vazkii/tinkerer/common/ThaumicTinkerer.java
/**
 * This class was created by <Vazkii>. It's distributed as
 * part of the ThaumicTinkerer Mod.
 *
 * ThaumicTinkerer is Open Source and distributed under a
 * Creative Commons Attribution-NonCommercial-ShareAlike 3.0 License
 * (http://creativecommons.org/licenses/by-nc-sa/3.0/deed.en_GB)
 *
 * ThaumicTinkerer is a Derivative Work on Thaumcraft 4.
 * Thaumcraft 4 (c) Azanor 2012
 * (http://www.minecraftforum.net/topic/1585216-)
 *
 * File Created @ [4 Sep 2013, 16:01:28 (GMT)]
 */
package vazkii.tinkerer.common;

import thaumcraft.common.CommonProxy;
import thaumcraft.common.Thaumcraft;
import vazkii.tinkerer.common.core.proxy.TTCommonProxy;
import vazkii.tinkerer.common.lib.LibMisc;
import cpw.mods.fml.common.Mod;
import cpw.mods.fml.common.Mod.EventHandler;
import cpw.mods.fml.common.Mod.Instance;
import cpw.mods.fml.common.SidedProxy;
import cpw.mods.fml.common.event.FMLInitializationEvent;
import cpw.mods.fml.common.event.FMLPostInitializationEvent;
import cpw.mods.fml.common.event.FMLPreInitializationEvent;

@Mod(modid = LibMisc.MOD_ID, name = LibMisc.MOD_NAME, version = LibMisc.VERSION, dependencies = LibMisc.DEPENDENCIES)
public class ThaumicTinkerer {

	@Instance(LibMisc.MOD_ID)
	public static ThaumicTinkerer instance;

	@SidedProxy(clientSide = LibMisc.CLIENT_PROXY, serverSide = LibMisc.COMMON_PROXY)
	public static TTCommonProxy proxy;

	public static CommonProxy tcProxy;

	@EventHandler
	public void preInit(FMLPreInitializationEvent event) {
		tcProxy = Thaumcraft.proxy;

		proxy.preInit(event);
	}

	@EventHandler
	public void init(FMLInitializationEvent event) {
		proxy.init(event);
	}

	@EventHandler
	public void postInit(FMLPostInitializationEvent event) {
		proxy.postInit(event);
	}
}

File: src/main/scala/com/nekokittygames/Thaumic/Tinkerer/common/ThaumicTinkerer.scala
package com.nekokittygames.Thaumic.Tinkerer.common;

import com.nekokittygames.Thaumic.Tinkerer.api.ThaumicTinkererAPI
import com.nekokittygames.Thaumic.Tinkerer.common.blocks.ModBlocks
import com.nekokittygames.Thaumic.Tinkerer.common.core.handler.PlayerHandler
import com.nekokittygames.Thaumic.Tinkerer.common.core.misc.{MobAspects, TTConfig}
import com.nekokittygames.Thaumic.Tinkerer.common.core.proxy.CommonProxy
import com.nekokittygames.Thaumic.Tinkerer.common.data.BoundJarNetworkManager
import com.nekokittygames.Thaumic.Tinkerer.common.items.ModItems
import com.nekokittygames.Thaumic.Tinkerer.common.items.baubles.ItemFoodTalisman
import com.nekokittygames.Thaumic.Tinkerer.common.libs.LibMisc
import com.nekokittygames.Thaumic.Tinkerer.common.research.{ModRecipes, ModResearch}
import net.minecraft.entity.Entity
import net.minecraft.nbt.{NBTTagList, NBTTagCompound}
import net.minecraftforge.common.util.Constants
import net.minecraftforge.fml.common.Mod.EventHandler
import net.minecraftforge.fml.common.event.FMLInterModComms.IMCMessage
import net.minecraftforge.fml.common.event.{FMLInterModComms, FMLInitializationEvent, FMLPostInitializationEvent, FMLPreInitializationEvent}
import net.minecraftforge.fml.common._
import org.apache.logging.log4j.Logger
import thaumcraft.api.aspects.Aspect
import thaumcraft.common
import thaumcraft.common.Thaumcraft
import scala.collection.JavaConversions._
import scala.collection.mutable

/**
 * Created by Katrina on 17/05/2015.
 */
@Mod(modid = LibMisc.MOD_ID,name = LibMisc.MOD_NAME,version = LibMisc.VERSION,dependencies = LibMisc.DEPENDENCIES,guiFactory = LibMisc.GUI_FACTORY,modLanguage = "scala")
object ThaumicTinkerer {
  final var logger: Logger = null


  @SidedProxy(clientSide = LibMisc.CLIENT_PROXY, serverSide = LibMisc.COMMON_PROXY)
  var proxy: CommonProxy = null


  var ThaumcraftProxy: common.CommonProxy = null

  @EventHandler
  def preInit(eventArgs: FMLPreInitializationEvent) = {
    TTConfig.init(eventArgs.getSuggestedConfigurationFile)
    ThaumcraftProxy = Thaumcraft.proxy
    logger = eventArgs.getModLog
    ModBlocks.registerBlocks(eventArgs)
    ModItems.registerItems(eventArgs)
    Loader.instance().getLoaderState

  }


  @EventHandler
  def init(eventArgs:FMLInitializationEvent): Unit =
  {
    proxy.init(eventArgs)
  }


  @EventHandler
  def imcCallback(event:FMLInterModComms.IMCEvent)= {

    for (message: IMCMessage <- event.getMessages) {
      if (message.key.equalsIgnoreCase(ThaumicTinkererAPI.FOOD_TALISMAN_IMC)) {
        ItemFoodTalisman.foodBlacklist += message.getStringValue
      }
      if (message.key.equalsIgnoreCase(ThaumicTinkererAPI.MOB_ASPECT_IMC)) {
        val cmp = message.getNBTValue
        try {
          val clazz: Class[_ <: Entity] = Class.forName(cmp.getString(ThaumicTinkererAPI.ENTITY_CLASS_NAME)).asInstanceOf[Class[_ <: Entity]]

          val aspects: scala.collection.mutable.MutableList[Aspect] = new mutable.MutableList[Aspect]()
          val aspectcmp: NBTTagList = message.getNBTValue.getTagList(ThaumicTinkererAPI.ASPECT_NAME, Constants.NBT.TAG_COMPOUND)
          for (i <- 0 until aspectcmp.tagCount()) {
            aspects.+=(Aspect.getAspect(aspectcmp.getCompoundTagAt(i).getString(ThaumicTinkererAPI.ASPECT_NAME)))
          }
          MobAspects.aspects.put(clazz, aspects.toArray)

        }
      }
    }
  }

  @EventHandler
  def Init(eventArgs: FMLInitializationEvent) = {
    ModBlocks.registerBlocksInventory()
    ModItems.registerInventoryItems()
    proxy.registerPacketHandlers()
    FMLCommonHandler.instance().bus().register(PlayerHandler)

    if (Loader.isModLoaded("Waila")) {
      FMLInterModComms.sendMessage("Waila", "register", "com.nekokittygames.Thaumic.Tinkerer.common.integration.Waila.callbackRegister");
    }

  }

  @EventHandler
  def postInit(eventArgs: FMLPostInitializationEvent) =
  {
    ModRecipes.registerRecipes()
    ModResearch.registerResearch()
    proxy.registerTileEntityRenders()
  }

}
